ql/termstructures/yield/regulatorytermstructure.hpp

```diff
@@ -0,0 +1,175 @@
+          from the official source.
+        - extrapolated forward is validated.
+        - rates on the cut-off point are checked against those
+          implied by the base curve.
+        - inspectors are tested against the base curve.
+        - incorrect input for cut-off point should raise an exception.
+        - observability against changes in the underlying term
+          structure and the additional components is checked.
+    */
+
+    class UltimateForwardTermStructure : public ZeroYieldStructure {
+      public:
+        UltimateForwardTermStructure(const Handle<YieldTermStructure>&,
+                                     const Handle<Quote>& lastLiquidForwardRate,
+                                     const Handle<Quote>& ultimateForwardRate,
+                                     Time firstSmoothingPoint,
+                                     Real alpha);
+        //! \name YieldTermStructure interface
+        //@{
+        DayCounter dayCounter() const;
+        Calendar calendar() const;
+        Natural settlementDays() const;
+        const Date& referenceDate() const;
+        Date maxDate() const;
+        Time maxTime() const;
+        //@}
+        //! \name Observer interface
+        //@{
+        void update();
+        //@}
+      protected:
+        //! returns the UFR extended zero yield rate
+        Rate zeroYieldImpl(Time) const;
+        //@}
+      private:
+        Handle<YieldTermStructure> originalCurve_;
+        Handle<Quote> llfr_;
+        Handle<Quote> ufr_;
+        Time fsp_;
+        Real alpha_;
+    };
+
+    // inline definitions
+
+    inline UltimateForwardTermStructure::UltimateForwardTermStructure(
+        const Handle<YieldTermStructure>& h,
+        const Handle<Quote>& lastLiquidForwardRate,
+        const Handle<Quote>& ultimateForwardRate,
+        Time firstSmoothingPoint,
+        Real alpha)
+    : originalCurve_(h), llfr_(lastLiquidForwardRate), ufr_(ultimateForwardRate),
+      fsp_(firstSmoothingPoint), alpha_(alpha) {
+        QL_REQUIRE(fsp_ > 0.0, "time to first smoothing point must be positive");
+        if (!originalCurve_.empty())
+            enableExtrapolation(originalCurve_->allowsExtrapolation());
+        registerWith(originalCurve_);
+        registerWith(llfr_);
+        registerWith(ufr_);
+    }
+
+    inline DayCounter UltimateForwardTermStructure::dayCounter() const {
+        return originalCurve_->dayCounter();
+    }
+
+    inline Calendar UltimateForwardTermStructure::calendar() const {
+        return originalCurve_->calendar();
+    }
+
+    inline Natural UltimateForwardTermStructure::settlementDays() const {
+        return originalCurve_->settlementDays();
+    }
+
+    inline const Date& UltimateForwardTermStructure::referenceDate() const {
+        return originalCurve_->referenceDate();
+    }
+
+    inline Date UltimateForwardTermStructure::maxDate() const { return originalCurve_->maxDate(); }
```

Comment:
Since the curve has a specified behavior past the smoothing point, I think its max date doesn't need to reflect that of the original curve and can be `Date::maxDate()`.  The same goes for the max time.  Unless there's some other specified limit?

Reply:
Date::maxDate() is a much better choice here. Thank you for pointing that out, I will change that.